lib/wasi/src/runtime/module_cache/shared.rs

```diff
@@ -0,0 +1,42 @@
+use std::{collections::HashMap, sync::Arc};
+
+use tokio::sync::RwLock;
+use wasmer::{Engine, Module};
+
+use crate::runtime::module_cache::{CacheError, CompiledModuleCache};
+
+/// A [`CompiledModuleCache`] based on a
+/// <code>[Arc]<[RwLock]<[HashMap]<[String], [Module]>>></code> that can be
+/// shared.
+#[derive(Debug, Default, Clone)]
+pub struct SharedCache {
+    modules: Arc<RwLock<HashMap<String, Module>>>,
+}
+
+impl SharedCache {
+    pub fn new() -> SharedCache {
+        SharedCache::default()
+    }
+
+    pub fn from_existing_modules(modules: Arc<RwLock<HashMap<String, Module>>>) -> Self {
+        SharedCache { modules }
+    }
+}
+
+#[async_trait::async_trait]
+impl CompiledModuleCache for SharedCache {
+    async fn load(&self, key: &str, _engine: &Engine) -> Result<Module, CacheError> {
+        let modules = self.modules.read().await;
```

Comment:
I don't think async here will be required if we use dashmap

Reply:
I've switched over to dashmap, but I'm planning to keep `ModuleCache` as an `async` trait.